### ISSUE TITLE
Example: large_deformation/compare_elastic_materials.py

### ISSUE THREAD

**Issue by Hailey-Liu** [sfepy/sfepy#568]
Hello guys,

I am trying to learn to use sfepy to simulation a large deformation tensile loading. In this example  `./examples/large deformation/compare_elastic_materials.py`, I found it perform by increasing the value of load on the top layer in each time step and calculate the displacement. If I want to perform in the other way that increasing the displacement of top layer in each time step and calculate the force, how can I do that? Thanks.

**Comment by rc** [comment]
Hi @HaominIsMe, check [1], where the time-dependent DIrichlet boundary conditions (ebcs) are used for the displacements.

[1] http://sfepy.org/doc-devel/examples/large_deformation/hyperelastic.html

**Comment by HaominIsMe** [comment]
Thanks @rc, it helps a lot. I could also get local stress and strain. Also, are there some methods to output the global stress along the tensile loading direction? I want to plot a stress-strain curve. If the loading scenario is load-control, the stress-strain curve is easy since the load value is pre-defined, and displacement of the top region can be measured. However, in the scenario of displacement-control, I couldn't find an example of plotting the stress-strain curve. Do you have any suggestions?

**Comment by HaominIsMe** [comment]
To get the global stress in x, y, z direction. I am trying to add a `post_process_hook`. But the calculated global stress is too high. Is there something wrong with my calculation?

```
def stress_strain(out, pb, state, extend=False):

	ev = pb.evaluate
	strain = ev('ev_cauchy_strain.2.Omega(u)', mode='el_avg')
	stress = ev('ev_cauchy_stress.2.Omega(solid.D, u)', mode='el_avg',
				copy_materials=False)

	out['cauchy_strain'] = Struct(name='output_data', mode='cell',data=strain, dofs=None)
	out['cauchy_stress'] = Struct(name='output_data', mode='cell',data=stress, dofs=None)

	min_z, max_z = pb.domain.get_mesh_bounding_box()[:, 2]
	box_size = max_z - min_z
	strain = 0.01 / box_size
	global_stress_xx = stress[:, 0, 0, 0].sum()
	global_stress_yy = stress[:, 0, 1, 0].sum()
	global_stress_zz = stress[:, 0, 2, 0].sum()
	# create new (or open existing) CSV file to hold the siffness value
	output_file = 'stiffness.csv'
	if os.path.isfile(output_file):
		f = open(output_file, 'ab')
		writer = csv.writer(f, encoding='utf-8')
	else:
		f = open(output_file, 'wb')
		writer = csv.writer(f, encoding='utf-8')
		writer.writerow(['stiffness'])
	writer.writerow([global_stress_xx/strain, global_stress_yy/strain, global_stress_zz/strain])
	f.close()

	return out
```

**Comment by rc** [comment]
Hi @HaominIsMe,

to get a force in some direction `n`, you need to compute the traction vector from stress `T_i = sigma_ij n_j`, which is per-unit-area (c.f. https://www.continuummechanics.org/tractionvector.html) and then integrate this over the cross-section area with the normal `n`. In SfePy, this could be done by using ev_cauchy_stress and ev_surface_integrate_mat terms, but there is no example readily available. As it is a common task, I will try modifying an existing elasticity example, or add a new one (not sure when I get to that though).

**Comment by HaominIsMe** [comment]
Hi @rc ,

Thanks for your reply. I was trying to understand how to use  ev_surface_integrate_mat term to get the force. But I am having difficulties in making it works. It would be very appreciated if an example could be given. Thanks.

**Comment by rc** [comment]
See #569 for an example in the linear elastic context. It is not that straightforward, and makes use of the fact, that the inner cross-section element faces are planar. In large deformation case, there might be some missing functionality - even in the linear case I worked around the fact, that there is no term to evaluate Cauchy stress on a surface. I chose this way instead of adding the term to show that a lot of things can be done by a user. Is is not difficult to add such a term, though.

Let me know whether the code in #569 is comprehensible, please.

**Comment by Hailey-Liu** [comment]
Hi @rc 

I check the code in #569. In the console, I got the output below:
> sfepy: surface load force: 4.0
> sfepy: middle section axial force: 3.999982514305424

This result is reasonable.
But when I modified the evaluation by changing the integration region to Right region or Left region. The results are different.
> sfepy: surface load force: 4.0
> sfepy: right section axial force: 4.638845227808088

> sfepy: surface load force: 4.0
> sfepy: left section axial force: 4.595471988788399

Is this reasonable?

Thanks

**Comment by rc** [comment]
I would guess that it is caused by a boundary effect and a coarse mesh. I have tried a uniform refinement:
```
python simple.py examples/linear_elasticity/linear_elastic_tractions.py -O refinement_level=1
...
sfepy: Middle section axial force: 3.999996871433383
sfepy: Right section axial force: 4.482621382017113
sfepy: Left section axial force: 4.437184423587806
```
and
```
python simple.py examples/linear_elasticity/linear_elastic_tractions.py -O refinement_level=2
...
sfepy: Middle section axial force: 3.9999989624606314
sfepy: Right section axial force: 4.331950108543268
sfepy: Left section axial force: 4.296347918999139
```
so it seems to converge. It might be interesting to plot the force accuracy along the axis.

**Comment by rc** [comment]
Increasing the polynomial order works better (note that you need to set the quadrature order to 2 * the approximation order in all the terms!):
```
    fields = {
        'displacement': ('real', 3, 'Omega', 2),
    }
```
leads to
```
sfepy: Middle section axial force: 4.000001586956472
sfepy: Right section axial force: 4.226081151912755
sfepy: Left section axial force: 4.190495914104636
```
and
```
    fields = {
        'displacement': ('real', 3, 'Omega', 3),
    }
```
to
```
sfepy: Middle section axial force: 4.000000065082226
sfepy: Right section axial force: 4.078301159067519
sfepy: Left section axial force: 4.06428364155672
```

**Comment by Hailey-Liu** [comment]
@rc Thanks for the helpful information. I will play around with it.

**Comment by rc** [comment]
FYI: I have updated #569 for the various cross-sections. Let me know if something is not clear in that example. If all is fine, I will merge it.

**Comment by rc** [comment]
I have just merged #569. I am closing issue, but feel free to reopen it, or ask/discuss further here.
